Follow-Up Fix Directory Open

nathanfranke opened the pull request:
Fixes #40234 
Related to #40151 

We should test this thoroughly before merging to make sure there aren't any other regressions.

CC @Xrayez @akien-mga 

```py
var fs = Directory.new()
print(fs.file_exists("res://icon.png")) # True
print(fs.dir_exists("res://")) # True

fs.open("test")
fs.list_dir_begin(true, true)
print(fs.get_next()) # test_file
```

Commit:
Follow-Up Fix Directory Open

nathanfranke commented:
Updated documentation

Xrayez commented on core/bind/core_bind.cpp:
```diff
@@ -1673,8 +1673,14 @@ int _Directory::get_current_drive() {
-	return d->change_dir(p_dir);
+	Error err = d->change_dir(p_dir);
+	
+	if(err != Error::OK) {
```
Scoping is not necessary, consistent with other codebase.
Also clang-format asks to add a space there. 🙂

```suggestion
	if (err != OK) {
```

Xrayez commented on core/bind/core_bind.cpp:
```diff
@@ -1673,8 +1673,14 @@ int _Directory::get_current_drive() {
+	}
+	dir_open = true;
+	
+	return Error::OK;
```
Same here.

```suggestion
	return OK;
```

Xrayez commented:
I've tested this on 3.2 branch, I haven't caught any errors. If there are, they can be easily fixed on the project side rather than plugins etc, so should be fine I guess. I'm also able to run unit tests with GUT now.

nathanfranke commented:
Thanks for the changes feedback. I always forget the space after `if` :)

akien-mga commented on core/bind/core_bind.cpp:
```diff
@@ -1673,8 +1673,14 @@ int _Directory::get_current_drive() {
-	ERR_FAIL_COND_V_MSG(!is_open(), ERR_UNCONFIGURED, "Directory must be opened before use.");
-	return d->change_dir(p_dir);
+	Error err = d->change_dir(p_dir);
+	
```
Trailing whitespace here and on line 1682.

I suggest setting up the pre-commit hooks locally to run clang-format when you commit: https://github.com/godotengine/godot/tree/master/misc/hooks

akien-mga commented on core/bind/core_bind.cpp:
```diff
@@ -1673,8 +1673,14 @@ int _Directory::get_current_drive() {
 }
 
 Error _Directory::change_dir(String p_dir) {
-	ERR_FAIL_COND_V_MSG(!is_open(), ERR_UNCONFIGURED, "Directory must be opened before use.");
```
I think the methods which don't check `is_open()` should keep checking that `d` is valid as they used to. Otherwise if the code ever changes to make `d` invalid, this is a null pointer dereference waiting to happen.

nathanfranke commented:
Fixed

I renamed the error so it wouldn't be confused with the `is_open` check

```c++
ERR_FAIL_COND_V_MSG(!d, ERR_UNCONFIGURED, "Directory is not configured properly.");
```

As of right now, I am not sure what "not configured properly" means since I don't think `d` can be `nullptr`

akien-mga commented:
> As of right now, I am not sure what "not configured properly" means since I don't think `d` can be `nullptr`

Well, the main question is: Can you guarantee that `DirAccess::create(DirAccess::ACCESS_RESOURCES);` would never return `nullptr`? It's a virtual method which has to be implemented by each platform port/driver, so it differs on Windows and Unix, might also differ on some console ports or other thirdparty ports, etc. Also differs when in a PCK (`DirAccessPack`).

So this kind of check is definitely useful in the wrapper code that relies on so many platform-specific implementations.

akien-mga commented:
```
In file included from ./core/cowdata.h:34,
                 from ./core/ustring.h:35,
                 from ./core/color.h:35,
                 from ./core/image.h:34,
                 from core/bind/core_bind.h:34,
                 from core/bind/core_bind.cpp:31:
core/bind/core_bind.cpp: In member function 'bool _Directory::file_exists(String)':
core/bind/core_bind.cpp:1715:26: error: enum constant in boolean context [-Werror=int-in-bool-context]
 1715 |  ERR_FAIL_COND_V_MSG(!d, ERR_UNCONFIGURED, "Directory is not configured properly.");
      |                          ^~~~~~~~~~~~~~~~
./core/error_macros.h:386:10: note: in definition of macro 'ERR_FAIL_COND_V_MSG'
  386 |   return m_retval;                                                                                                                            \
      |          ^~~~~~~~
core/bind/core_bind.cpp: In member function 'bool _Directory::dir_exists(String)':
core/bind/core_bind.cpp:1724:26: error: enum constant in boolean context [-Werror=int-in-bool-context]
 1724 |  ERR_FAIL_COND_V_MSG(!d, ERR_UNCONFIGURED, "Directory is not configured properly.");
      |                          ^~~~~~~~~~~~~~~~
./core/error_macros.h:386:10: note: in definition of macro 'ERR_FAIL_COND_V_MSG'
  386 |   return m_retval;                                                                                                                            \
      |          ^~~~~~~~
```

nathanfranke commented:
Apologies! `file_exists` and `dir_exists` return a `bool` and not an `Error`

akien-mga commented:
Thanks!

akien-mga commented:
Cherry-picked for 3.2.3.
